### PR TITLE
feat: correlation support

### PR DESCRIPTION
**Pull request by huaxingao**
## Which issue does this PR close?

Supports `CORR`
The implementation mostly is the same as the DataFusion's implementation. The reason
we have our own implementation is that DataFusion has UInt64 for state_field count,
while Spark has Double for count. Also adding null_on_divide_by_zero
to be consistent with Spark's implementation.
Closes #.

### CODE REVIEW

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala [hunk 1]
```diff
@@ -547,6 +547,26 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
...
+          corrBuilder.setNullOnDivideByZero(nullOnDivideByZero)
+          corrBuilder.setDatatype(dataType.get)
+
+          Some(
+            ExprOuterClass.AggExpr
+              .newBuilder()
+              .setCorrelation(corrBuilder)
+              .build())
+        } else {
+          None
```
Comment (kazuyukitanimura) [inline spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:568]: Should we add `withInfo()`?

File: spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala [hunk 2]
```diff
@@ -1212,6 +1212,157 @@ class CometAggregateSuite extends CometTestBase with AdaptiveSparkPlanHelper {
...
+                withSQLConf(
+                  "spark.sql.legacy.statisticalAggregate" -> nullOnDivideByZero.toString) {
+                  val table = "test"
+                  withTable(table) {
+                    sql(
+                      s"create table $table(col1 double, col2 double, col3 double) using parquet")
+                    sql(s"insert into $table values(1, 4, 1), (2, 5, 1), (3, 6, 2)")
+                    val expectedNumOfCometAggregates = 2
+
+                    sql("SELECT corr(col1, col2) FROM test GROUP BY col3").show
```
Comment (kazuyukitanimura) [inline spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala:1232]: nit: we can reuse s`$table`

**Review by kazuyukitanimura — APPROVED** [review]

File: spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala (hunk 1 above)
Comment (huaxingao) [reply spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:569]: Yes, I forgot this. Added.

File: spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala (hunk 2 above)
Comment (huaxingao) [reply spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala]: Done.

**Comment by huaxingao** [comment]
cc @andygrove @viirya

File: spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala [hunk 3]
```diff
@@ -1212,6 +1212,157 @@ class CometAggregateSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     }
   }
 
+  test("correlation") {
+    withSQLConf(CometConf.COMET_EXEC_SHUFFLE_ENABLED.key -> "true") {
+      Seq(false).foreach { cometColumnShuffleEnabled =>
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala:1217]: Is it intentional that we only test for `cometColumnShuffleEnabled==false`? I have similar questions for `parquet.enable.dictionary` and `nullOnDivideByZero`, where we also iterate over a single boolean and don't test both for true and false.

Comment (huaxingao) [reply spark/src/test/scala/org/apache/comet/exec/CometAggregateSuite.scala:1217]: Should be both `true` and `false`. I removed one of them so it's easier for me to debug. Forgot to put back.

**Review by andygrove — APPROVED** [review]
LGTM

**Comment by viirya** [comment]
Merged. Thanks @huaxingao @andygrove @kazuyukitanimura

**Pull request merged**